cmd/snap-confine/udev-support.c

```diff
@@ -18,290 +18,355 @@
-#include <sched.h>
+#include <fcntl.h>
 #include <string.h>
 #include <sys/stat.h>
+#include <sys/sysmacros.h>
 #include <sys/types.h>
-#include <sys/wait.h>
 #include <unistd.h>
 
+#include <libudev.h>
+
+#include "../libsnap-confine-private/cleanup-funcs.h"
 #include "../libsnap-confine-private/snap.h"
 #include "../libsnap-confine-private/string-utils.h"
+#include "../libsnap-confine-private/cgroup-support.h"
 #include "../libsnap-confine-private/utils.h"
 #include "udev-support.h"
 
-static void
-_run_snappy_app_dev_add_majmin(struct snappy_udev *udev_s,
-			       const char *path, unsigned major, unsigned minor)
+/* Allow access to common devices. */
+static void sc_udev_allow_common(int devices_allow_fd)
 {
-	int status = 0;
-	pid_t pid = fork();
-	if (pid < 0) {
-		die("cannot fork support process for device cgroup assignment");
-	}
-	if (pid == 0) {
-		uid_t real_uid, effective_uid, saved_uid;
-		if (getresuid(&real_uid, &effective_uid, &saved_uid) != 0)
-			die("cannot get real, effective and saved user IDs");
-		// can't update the cgroup unless the real_uid is 0, euid as
-		// 0 is not enough
-		if (real_uid != 0 && effective_uid == 0)
-			if (setuid(0) != 0)
-				die("cannot set user ID to zero");
-		char buf[64] = { 0 };
-		// pass snappy-add-dev an empty environment so the
-		// user-controlled environment can't be used to subvert
-		// snappy-add-dev
-		char *env[] = { NULL };
-		if (minor == UINT_MAX) {
-			sc_must_snprintf(buf, sizeof(buf), "%u:*", major);
-		} else {
-			sc_must_snprintf(buf, sizeof(buf), "%u:%u", major,
-					 minor);
-		}
-		debug("running snap-device-helper add %s %s %s",
-		      udev_s->tagname, path, buf);
-		// This code runs inside the core snap. We have two paths
-		// for the udev helper.
-		//
-		// First try new "snap-device-helper" path first but
-		// when running against an older core snap fallback to
-		// the old name.
-		if (access("/usr/lib/snapd/snap-device-helper", X_OK) == 0)
-			execle("/usr/lib/snapd/snap-device-helper",
-			       "/usr/lib/snapd/snap-device-helper", "add",
-			       udev_s->tagname, path, buf, NULL, env);
-		else if (access("/usr/libexec/snapd/snap-device-helper", X_OK) == 0)
-			execle("/usr/libexec/snapd/snap-device-helper",
-			       "/usr/libexec/snapd/snap-device-helper", "add",
-			       udev_s->tagname, path, buf, NULL, env);
-		else
-			execle("/lib/udev/snappy-app-dev",
-			       "/lib/udev/snappy-app-dev", "add",
-			       udev_s->tagname, path, buf, NULL, env);
-		die("execl failed");
-	}
-	if (waitpid(pid, &status, 0) < 0)
-		die("waitpid failed");
-	if (WIFEXITED(status) && WEXITSTATUS(status) != 0)
-		die("child exited with status %i", WEXITSTATUS(status));
-	else if (WIFSIGNALED(status))
-		die("child died with signal %i", WTERMSIG(status));
+	/* The devices we add here have static number allocation.
+	 * https://www.kernel.org/doc/html/v4.11/admin-guide/devices.html */
+	dprintf(devices_allow_fd, "c 1:3 rwm");	// /dev/null
```

Comment:
Looked at `snap-device-helper` and back here, I think the current code is not entirely equivalent/ The `s-d-h` runs with `set -e`, and the C code that ran it called `die()` on any error. We probably need a wrapper that does more less this:

```
void dprintf_or_die(int fd, const char *format, ...) {
    va_list args;
    va_start(args, format);
    if (vdprintf(fd, format, args) == -1) {
        die("failed to write");
    }
    va_end(list);
}
```

Reply:
Done